packages/sanity/src/desk/panes/documentList/useDocumentList.ts

```diff
@@ -1,111 +1,211 @@
-  items: DocumentListPaneItem[] | null
+  isSearchReady: boolean
+  items: DocumentListPaneItem[]
+  onListChange: () => void
   onRetry?: (event: unknown) => void
 }
 
+const INITIAL_QUERY_RESULTS: QueryResult = {
+  result: null,
+  error: null,
+}
+
 /**
  * @internal
  */
 export function useDocumentList(opts: UseDocumentListOpts): DocumentListState {
-  const {apiVersion, filter, params, sortOrder} = opts
+  const {filter, params: paramsProp, sortOrder, searchQuery} = opts
   const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
-  const [fullList, setFullList] = useState(false)
-  const fullListRef = useRef(fullList)
-  const [result, setResult] = useState<QueryResult | null>(null)
-  const error = result?.error || null
-  const isLoading = result?.loading || result === null
-  const onRetry = result?.onRetry
-  const documents = result?.result?.documents
+  const schema = useSchema()
+
+  const [resultState, setResult] = useState<QueryResult>(INITIAL_STATE)
+  const {onRetry, error, result, loading, isInitialLoading} = resultState
+
+  const documents = result?.documents
+
+  // Filter out published documents that have drafts to avoid duplicates in the list.
   const items = useMemo(
-    () => (documents ? removePublishedWithDrafts(documents) : null),
+    () => (documents ? removePublishedWithDrafts(documents) : EMPTY_ARRAY),
     [documents]
   )
 
-  const query = useMemo(() => {
-    const extendedProjection = sortOrder?.extendedProjection
-    const projectionFields = ['_id', '_type']
-    const finalProjection = projectionFields.join(',')
-    const sortBy = sortOrder?.by || []
-    const limit = fullList ? FULL_LIST_LIMIT : PARTIAL_PAGE_LIMIT
-    const sort = sortBy.length > 0 ? sortBy : DEFAULT_ORDERING.by
-    const order = toOrderClause(sort)
-
-    if (extendedProjection) {
-      const firstProjection = projectionFields.concat(extendedProjection).join(',')
-      return [
-        `*[${filter}] {${firstProjection}}`,
-        `order(${order}) [0...${limit}]`,
-        `{${finalProjection}}`,
-      ].join('|')
-    }
+  // A state variable to keep track of whether we are currently lazy loading the list.
+  // This is used to determine whether we should show the loading spinner at the bottom of the list.
+  const [isLazyLoading, setIsLazyLoading] = useState<boolean>(false)
+
+  // A state to keep track of whether we have fetched the full list of documents.
+  const [hasFullList, setHasFullList] = useState<boolean>(false)
+
+  // A state to keep track of whether we should fetch the full list of documents.
+  const [shouldFetchFullList, setShouldFetchFullList] = useState<boolean>(false)
+
+  // Get the type name from the filter, if it is a simple type filter.
+  const typeNameFromFilter = useMemo(
+    () => getTypeNameFromSingleTypeFilter(filter, paramsProp),
+    [filter, paramsProp]
+  )
 
-    return `*[${filter}]|order(${order})[0...${limit}]{${finalProjection}}`
-  }, [filter, fullList, sortOrder])
+  const isLoading =
```

Comment:
Really minor – a clarifying comment here would be great

i.e. why is `isLoading = true` when `Boolean(loading && result === null && !error)`? Logically I'd assume that `isLoading` would be true even when `loading` is true, even with a `null` result. Perhaps I'm getting thrown off by the variable name itself.

Additionally – is `isInitialLoading` still being used?